packages/connect/src/connector/extension.ts

```diff
@@ -0,0 +1,133 @@
+  origin: "substrate-connect-client"
+} & infer V
+  ? Omit<V, "chainId">
+  : unknown
+type HeaderlessToExtension = HeaderlessToExtensionGeneric<ToExtension>
+
+const listeners = new Map<string, (msg: ToApplication) => void>()
+window.addEventListener("message", ({ data }: MessageEvent<ToApplication>) => {
+  if (data?.origin !== "substrate-connect-extension") return
+  listeners.get(data.chainId)?.(data)
+})
+
+const listenToExtension = (
+  chainId: string,
+  callback: (msg: ToApplication) => void,
+): (() => void) => {
+  listeners.set(chainId, callback)
+  return () => {
+    listeners.delete(chainId)
+  }
+}
+
+function getRandomChainId(): string {
+  const arr = new BigUint64Array(2)
+  // It can only be used from the browser, so this is fine.
+  crypto.getRandomValues(arr)
+  const result = (arr[1] << BigInt(64)) | arr[0]
+  return result.toString(36)
+}
+
+export const getConnectorClient = (): SubstrateConnector => {
+  const chains = new Map<Chain, string>()
+
+  const internalAddChain = async (
+    isWellKnown: boolean,
+    chainSpecOrWellKnownName: string,
+    jsonRpcCallback?: JsonRpcCallback,
+    potentialRelayChainIds = [] as string[],
+  ): Promise<Chain> => {
+    const chainId = getRandomChainId()
+    const postToExtension = (msg: HeaderlessToExtension) => {
+      window.postMessage(
+        {
+          ...msg,
+          origin: "substrate-connect-client",
+          chainId,
+        },
+        "*",
+      )
+    }
+
+    await new Promise<void>((res, rej) => {
+      const stopListeningToExtension = listenToExtension(chainId, (msg) => {
+        stopListeningToExtension()
+        if (msg.type === "chain-ready") return res()
+        rej(new Error("There was an error creating the smoldot chain."))
+      })
+
+      postToExtension(
+        isWellKnown
+          ? {
+              type: "add-well-known-chain",
+              chainName: chainSpecOrWellKnownName,
+            }
+          : {
+              type: "add-chain",
+              chainSpec: chainSpecOrWellKnownName,
+              potentialRelayChainIds,
+            },
+      )
+    })
+
+    const chainFn =
+      <Args extends Array<unknown>>(fn: (...args: Args) => void) =>
+      (...args: Args) => {
+        if (crashError) throw crashError
+        if (!chains.has(chain)) throw new AlreadyDestroyedError()
+        fn(...args)
+      }
+
+    const chain: Chain = {
+      sendJsonRpc: chainFn((jsonRpcMessage) => {
+        if (!jsonRpcCallback) throw new JsonRpcDisabledError()
+        postToExtension({ type: "rpc", jsonRpcMessage })
+      }),
+      remove: chainFn(() => {
+        stopListeningToExtension()
+        chains.delete(chain)
+        postToExtension({ type: "remove-chain" })
+      }),
+    }
+    chains.set(chain, chainId)
+
+    let crashError: CrashError | null = null
+    const stopListeningToExtension = listenToExtension(chainId, (msg) => {
+      if (msg.type !== "rpc" || !jsonRpcCallback) {
```

Comment:
I really don't think this is necessary. The web page assumes that the extension works, it's pointless to try to gracefully handle bugs in the extension.
In my opinion `crashError` can be removed.

Reply:
I mean, it doesn't hurt either and OTOH it helps TS and eslint be happy so that I don't leave unhandled branches behind... Also, if one day we had an issue with the Extension, this would provide us with more context of what happened.